sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubListener.PartitionProcessor.cs

```diff
@@ -0,0 +1,356 @@
+                                _cachedEventsBackgroundTaskCts = CancellationTokenSource.CreateLinkedTokenSource(_cts.Token);
+                                _cachedEventsBackgroundTask = MonitorCachedEvents(_cachedEventsBackgroundTaskCts.Token);
+                            }
+                        }
+                        finally
+						{
+							_cachedEventsGuard.Release();
+						}
+					}
+					else
+					{
+						UpdateCheckpointContext(events, context);
+						await TriggerExecute(events, context, _cts.Token).ConfigureAwait(false);
+						eventToCheckpoint = events.LastOrDefault();
+					}
+
+					// If total events is less than the batch size, leave them in the stored events list
+					// and wait to send until we receive enough events or total max wait time has passed.
+				}
+
+				// Checkpoint if we processed any events.
+				// Don't checkpoint if no events. This can reset the sequence counter to 0.
+				// Note: we intentionally checkpoint the batch regardless of function
+				// success/failure. EventHub doesn't support any sort "poison event" model,
+				// so that is the responsibility of the user's function currently. E.g.
+				// the function should have try/catch handling around all event processing
+				// code, and capture/log/persist failed events, since they won't be retried.
+				if (eventToCheckpoint != null)
+				{
+					await CheckpointAsync(eventToCheckpoint, context).ConfigureAwait(false);
+				}
+			}
+
+			private async Task TriggerExecute(EventData[] events, EventProcessorHostPartition context, CancellationToken cancellationToken)
+			{
+				var triggerInput = new EventHubTriggerInput
+				{
+					Events = events,
+					ProcessorPartition = context
+				};
+
+				// Batch dispatch
+				TriggeredFunctionData input = new()
+				{
+					TriggerValue = triggerInput,
+					TriggerDetails = triggerInput.GetTriggerDetails(context)
+				};
+
+				await _executor.TryExecuteAsync(input, cancellationToken).ConfigureAwait(false);
+
+				if (_cachedEventsBackgroundTaskCts != null)
+				{
+					// If there is a background timer task, cancel it and dispose of the cancellation token.
+					_cachedEventsBackgroundTaskCts.Cancel();
+					_cachedEventsBackgroundTaskCts.Dispose();
+					_cachedEventsBackgroundTaskCts = null;
+				}
+			}
+
+			private async Task MonitorCachedEvents(CancellationToken cancellationToken)
+			{
+				try
+				{
+					_currentCycle = ValueStopwatch.StartNew();
+
+					while (_currentCycle.GetElapsedTime() < _maxWaitTime && !cancellationToken.IsCancellationRequested)
+					{
+						var remainingTime = RemainingTime(_currentCycle.GetElapsedTime());
+						await Task.Delay(remainingTime, cancellationToken).ConfigureAwait(false);
+					}
+
+                    if (!_cachedEventsGuard.Wait(0, cancellationToken))
+                    {
+                        await _cachedEventsGuard.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
We need to release this in a `finally` block - but only if we actually acquire it.   It is possible for us to hit an exception that would cause `finally` to run before the semaphore is required.   We typically create a boolean var that we set to `true` right after the semaphore block to indicate that we need to release it and check that in `finally`.